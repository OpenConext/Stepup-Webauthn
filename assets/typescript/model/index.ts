export * from './application';
export * from './webauthn';
