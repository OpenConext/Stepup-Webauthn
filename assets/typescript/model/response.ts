export type ServiceResponseStatus = 'deviceNotSupported' |
  'noRegistrationRequired' |
  'noAuthenticationRequired' |
  'missingAttestationStatement' |
  'invalid' |
  'error' |
  'ok';

export interface ServerResponse {
  status: ServiceResponseStatus;
  error_code?: string;
}
