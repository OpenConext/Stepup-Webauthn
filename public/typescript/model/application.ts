
export type TranslationString = string;

export enum ApplicationEvent {
  // Generic
  NOT_SUPPORTED,
  SERIALIZE_PUBLIC_KEY_CREDENTIALS,
  PUBLIC_KEY_CREDENTIALS_SERIALIZED,
  SENDING_PUBLIC_KEY_CREDENTIALS,
  RECEIVED_SERVER_RESPONSE,

  // Registration
  DESERIALIZE_ATTESTATION_RESPONSE_OPTIONS,
  ATTESTATION_RESPONSE_OPTIONS_DE_SERIALIZED,
  REQUEST_USER_FOR_ASSERTION,

  // Authentication
  DESERIALIZE_ASSERTION_RESPONSE_OPTIONS,
  ASSERTION_RESPONSE_OPTIONS_DE_SERIALIZED,
  REQUEST_USER_FOR_ATTESTATION,
  PUBLIC_KEY_CREDENTIALS,

  // Error
  UNSUPPORTED_PUBLIC_KEY_CREDENTIALS,
  ERROR,
  UNSUPPORTED_CREDENTIAL_TYPE,
}

export interface RequestInformation {
  sari: string;
  hostname: string;
  requestId: string;
  userAgent: string;
  ipAddress: string;
  supportEmail: string;
}

export interface ErrorWithMail {
  code: string;
  error: string;
  timestamp: string;
  showMailTo: true;
  showRetry: boolean;
}

export interface ErrorWithoutMail {
  code: string;
  timestamp: string;
  showMailTo: false;
  showRetry: boolean;
}

export function errorWithMailTo(error: ErrorInformation): error is ErrorWithMail {
  return error.showMailTo;
}

export type ErrorInformation = ErrorWithMail | ErrorWithoutMail;

export interface ApplicationState {
  requestInformation: RequestInformation;
  errorInfo: ErrorInformation | null;
  message: TranslationString;

  // Debug values, not used in application.
  clientDataJSON?: string;
}

export interface ApplicationAction {
  type: ApplicationEvent;
  value: unknown;
  timestamp: string;
}

export type FireApplicationEvent = (event: ApplicationEvent) => (value: unknown) => void;
