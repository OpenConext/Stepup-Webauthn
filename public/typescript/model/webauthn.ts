export interface SerializedPublicKeyCredentialUserEntity extends Omit<PublicKeyCredentialUserEntity, 'id'> {
  id: string;
}

export interface SerializedPublicKeyCredentialDescriptor extends Omit<PublicKeyCredentialDescriptor, 'id'> {
  id: string;
}

export interface SerializedPublicKeyCredentialCreationOptions extends Omit<PublicKeyCredentialCreationOptions, 'challenge' | 'excludeCredentials' | 'user'> {
  challenge: string;
  user: SerializedPublicKeyCredentialUserEntity;
  excludeCredentials?: SerializedPublicKeyCredentialDescriptor[];
}

export interface SerializedPublicKeyCredentialRequestOptions extends Omit<PublicKeyCredentialRequestOptions, 'challenge' | 'allowCredentials'> {
  challenge: string;
  allowCredentials?: SerializedPublicKeyCredentialDescriptor[];
}

export interface SerializedAuthenticatorAttestationResponse extends Omit<AuthenticatorAttestationResponse, 'clientDataJSON' | 'attestationObject'> {
  clientDataJSON: string;
  attestationObject: string | undefined;
}

export interface SerializedAuthenticatorAssertionResponse extends Omit<AuthenticatorAssertionResponse, 'clientDataJSON' | 'userHandle' | 'signature' | 'authenticatorData'> {
  clientDataJSON: string;
  userHandle: string | undefined;
  signature: string;
  authenticatorData: string;
}

export type SerializedAuthenticatorResponse =
  SerializedAuthenticatorAttestationResponse
  | SerializedAuthenticatorAssertionResponse;

export interface SerializedPublicKeyCredential extends Omit<PublicKeyCredential, 'rawId' | 'response'> {
  rawId: string;
  response: SerializedAuthenticatorResponse;
}
