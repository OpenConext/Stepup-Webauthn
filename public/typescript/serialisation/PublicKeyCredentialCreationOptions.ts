import { SerializedPublicKeyCredentialCreationOptions, SerializedPublicKeyCredentialRequestOptions } from '../model';
import {
  base64UrlSafeToUInt8,
  base64UrlSafeToUInt8Id,
  optionalBase64UrlSafeToUInt8Ids,
  removeEmptyAndUndefined,
} from './common';

export const deSerializedPublicKeyCredentialCreationOptions: (options: SerializedPublicKeyCredentialCreationOptions) => PublicKeyCredentialCreationOptions =
  ({ rp, user, challenge, extensions, attestation, authenticatorSelection, timeout, pubKeyCredParams, excludeCredentials }) =>
    removeEmptyAndUndefined(({
      rp,
      user: base64UrlSafeToUInt8Id(user),
      challenge: base64UrlSafeToUInt8(challenge),
      pubKeyCredParams,
      timeout,
      excludeCredentials: optionalBase64UrlSafeToUInt8Ids(excludeCredentials),
      authenticatorSelection,
      attestation,
      extensions,
    }));

export const deSerializedPublicKeyCredentialRequestOptions: (options: SerializedPublicKeyCredentialRequestOptions) => PublicKeyCredentialRequestOptions =
  ({ challenge, extensions, timeout, allowCredentials, rpId, userVerification }) =>
    removeEmptyAndUndefined(({
      rpId,
      challenge: base64UrlSafeToUInt8(challenge),
      timeout,
      allowCredentials: optionalBase64UrlSafeToUInt8Ids(allowCredentials),
      userVerification,
      extensions,
    }));
