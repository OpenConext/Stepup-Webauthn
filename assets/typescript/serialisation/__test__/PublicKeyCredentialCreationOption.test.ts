import { SerializedPublicKeyCredentialCreationOptions } from '../../model';
import { deSerializedPublicKeyCredentialCreationOptions } from '../PublicKeyCredentialCreationOptions';

it('deSerializedPublicKeyCredentialCreationOptions', () => {
  const publicKeyOptions: SerializedPublicKeyCredentialCreationOptions = {
    rp: { name: 'My Super Secured Application', id: 'foo.example.com' },
    pubKeyCredParams: [{ type: 'public-key', alg: -7 }],
    challenge: '237glVYXXQoFNJCF_7fgCFwKneTN5QsWWR9hXT41rTY',
    attestation: 'none',
    user: {
      name: '@cypher-Angel-3000',
      id: 'MTIzZTQ1NjctZTg5Yi0xMmQzLWE0NTYtNDI2NjU1NDQwMDAw',
      displayName: 'Mighty Mike',
    },
    authenticatorSelection: { requireResidentKey: false, userVerification: 'preferred' },
    excludeCredentials: [{ type: 'public-key', id: 'QUJDREVGR0g=' }],
    extensions: { loc: true },
    timeout: 20000,
  };
  expect(deSerializedPublicKeyCredentialCreationOptions(publicKeyOptions)).toMatchSnapshot();
});

it('deSerializedPublicKeyCredentialCreationOptions should remove empty options', () => {
  const publicKeyOptions: SerializedPublicKeyCredentialCreationOptions = {
    rp: { name: 'My Super Secured Application', id: 'foo.example.com' },
    pubKeyCredParams: [],
    challenge: '237glVYXXQoFNJCF_7fgCFwKneTN5QsWWR9hXT41rTY',
    attestation: 'none',
    user: {
      name: '@cypher-Angel-3000',
      id: 'MTIzZTQ1NjctZTg5Yi0xMmQzLWE0NTYtNDI2NjU1NDQwMDAw',
      displayName: 'Mighty Mike',
    },
    authenticatorSelection: { requireResidentKey: false, userVerification: 'preferred' },
    excludeCredentials: undefined,
    extensions: { loc: true },
    timeout: 20000,
  };
  expect(deSerializedPublicKeyCredentialCreationOptions(publicKeyOptions)).toMatchSnapshot();
});
