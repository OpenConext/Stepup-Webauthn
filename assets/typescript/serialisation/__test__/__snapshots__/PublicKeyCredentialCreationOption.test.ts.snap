// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deSerializedPublicKeyCredentialCreationOptions 1`] = `
Object {
  "attestation": "none",
  "authenticatorSelection": Object {
    "requireResidentKey": false,
    "userVerification": "preferred",
  },
  "challenge": ArrayBuffer [],
  "excludeCredentials": Array [
    Object {
      "id": ArrayBuffer [],
      "type": "public-key",
    },
  ],
  "extensions": Object {
    "loc": true,
  },
  "pubKeyCredParams": Array [
    Object {
      "alg": -7,
      "type": "public-key",
    },
  ],
  "rp": Object {
    "id": "foo.example.com",
    "name": "My Super Secured Application",
  },
  "timeout": 20000,
  "user": Object {
    "displayName": "Mighty Mike",
    "id": ArrayBuffer [],
    "name": "@cypher-Angel-3000",
  },
}
`;

exports[`deSerializedPublicKeyCredentialCreationOptions should remove empty options 1`] = `
Object {
  "attestation": "none",
  "authenticatorSelection": Object {
    "requireResidentKey": false,
    "userVerification": "preferred",
  },
  "challenge": ArrayBuffer [],
  "extensions": Object {
    "loc": true,
  },
  "rp": Object {
    "id": "foo.example.com",
    "name": "My Super Secured Application",
  },
  "timeout": 20000,
  "user": Object {
    "displayName": "Mighty Mike",
    "id": ArrayBuffer [],
    "name": "@cypher-Angel-3000",
  },
}
`;
